### PR TITLE
Improve text conversion for copying and highlighting from PDF documents

### PR DESCRIPTION
**Pull request by mvasilak**
* Improves copied text from PDF documents by glueing hyphenated words in new lines and removing new lines.
* Improves annotation text by by glueing hyphenated before certain new lines are removed.

Improvement mentioned in this [forum discussion](https://forums.zotero.org/discussion/109222/ios-pdf-reader-improvement-request)
With some info from from https://github.com/zotero/zotero-ios/issues/308

~- [ ] Also add database migration for annotation text if needed.~

### CODE REVIEW

File: Zotero/Controllers/TextConverter.swift [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        }
+        return newString
+    }
+
+    static func removeHyphenAndNewlinePairs(from string: String) -> String {
+        guard let expression = hyphenAndNewlineExpression else { return string }
+        return replaceMatches(of: expression, with: "", in: string)
+    }
+
+    static func replaceNewlineOptionallySurroundedByWhitespaceExcludingEndingSentencesExpressionWithSingleSpace(from string: String) -> String {
```
Comment (michalrentka) [inline Zotero/Controllers/TextConverter.swift:59]: Probably one of the longest names I've seen in Swift

**Review by michalrentka — APPROVED** [review]

**Comment by michalrentka** [comment]
I probably wouldn't do db migration just so that all of your highlight annotations don't suddenly change text (even if it's for the better) values. @dstillman do you agree?

**Comment by dstillman** [comment]
How would migration even work? We can't actually know what annotation text was created by the iOS app vs. the desktop app, right? It might be one thing to automatically update annotation text when a PDF was opened if we knew that the previous text was created on iOS, but otherwise I don't think it's really an option.

**Comment by mvasilak** [comment]
@dstillman the transformation would just affect annotations created on iOS, where it would glue hyphened words separated by newlines, and then would remove any extraneous newlines (it keeps those newlines after periods, as they might be paragraphs). Since the desktop app already does a superset of these actions, text of highlight annotations created with it, won't change.

Of course, it still might not worth doing.

**Comment by dstillman** [comment]
You're saying it wouldn't reprocess the text from the PDF but rather adjust a specific pattern (hyphen plus newline) in the stored annotation text itself?

**Comment by mvasilak** [comment]
Correct, the migration would act on the stored highlight annotation text. However, that means that it would act only on those synced currently. And if the user is still using an older client in another device, it might still create new ones with the old transformer. Hmmm, maybe it makes more sense to only update those documents that are opened, and keep track of this per document? Or do nothing as @michalrentka suggested?

**Comment by mvasilak** [comment]
Let's do nothing then for the time being, and we can see if we get relevant user feedback. To be honest, the best "migration" in this case might be to let users trigger such a transformation centrally in the backend database, not the local iOS database.

**Comment by mvasilak** [comment]
@dstillman you agree with doing nothing for the time being migration-wise ☝️

**Comment by dstillman** [comment]
Sure.

**Pull request merged**